### ISSUE TITLE
Feature idea: deduplication of requests

### ISSUE THREAD

**Issue by jamesplease** [FormidableLabs/urql#30]
By default, Apollo dedupes requests that contain the same root queries using [`apollo-link-dedup`](https://www.apollographql.com/docs/link/links/dedup.html).

urql already generates [a key for the response cache](https://github.com/FormidableLabs/urql/blob/39d81e2e09de08d3f6e5c4ad9e0d0287efecd555/src/modules/client.ts#L122), and that same key could be used for request deduplication. When response caching and request deduplication use the same underlying key, then your deduplication is as granular as your cache, which is cool 😎 

So, for instance, right now `urql` could deduplicate the _exact same query_, but if #16 is ever implemented, then it could dedupe requests like the following:

```
{
  QueryA
}
```

```
{
  QueryA
  QueryB
}
```

I'm working on a lib called `fetch-dedupe` that abstracts `apollo-link-dedupe`'s functionality into a generic-purpose lib, but for now [here's the code](https://github.com/jmeas/react-request/blob/57a404bf8f6c660004c9f9a084bffaeaef3b436b/src/fetch-dedupe.js).

---

If you were to copy pasta that file into urql, then the diff is actually pretty small to implement request deduplication:

```diff
- fetch(this.url, {
+ fetchDedupe(this.url, {
  body,
  headers: { 'Content-Type': 'application/json' },
  method: 'POST',
  ...fetchOptions,
- })
+ }, {
+   requestKey: hash,
+   responseType: 'json'
+ });
- .then(res => res.json())
```

although you might also want to add a new config that lets folks skip the deduplication, similar to Apollo's.

I'd PR a preview but I've never written TypeScript :v:

> Note: some modifications would likely need to be made to `fetchDedupe` to support sub-request body deduplication as described in #16

**Comment by kenwheeler** [comment]
This looks great. If you wanna submit a PR with failing types I could fix it up

**Comment by kitten** [comment]
Implemented on the `next` branch for v0.3 https://github.com/FormidableLabs/urql/blob/next/src/modules/dedup-exchange.ts
